### PR TITLE
Prevent exception in PCDReader for misformed PCD files

### PR DESCRIPTION
**Pull request by mvieth**

### CODE REVIEW

**Comment by larshg** [comment]
Is it manually edited files, since its incomplete? Or a failure at writing a file somehow?

Maybe print the index number if one wants to find it (not sure if header is a fixed line count as well) ?

**Comment by mvieth** [comment]
> Is it manually edited files, since its incomplete? Or a failure at writing a file somehow?

The files that are written with the "official" PCDWriter should always be correct. Yes, manual editing could create such a problem, or any kind of corruption.

> Maybe print the index number if one wants to find it (not sure if header is a fixed line count as well) ?

I don't see an easy way to print the actual line number of the PCD file, but instead the point index is printed now. That at least gives a hint where the problem is.

**Review by larshg — APPROVED** [review]

**Review by koide3 — APPROVED** [review]

File: io/src/pcd_io.cpp [hunk 1]
```diff
@@ -424,6 +424,10 @@ pcl::PCDReader::readBodyASCII (std::istream &fs, pcl::PCLPointCloud2 &cloud, int
 {
   // Get the number of points the cloud should have
   unsigned int nr_points = cloud.width * cloud.height;
+  // The number of elements each line/point should have
+  const unsigned int elems_per_line = std::accumulate (cloud.fields.begin (), cloud.fields.end (), 0u,
```
Comment (kunaltyagi) [inline io/src/pcd_io.cpp:428]: `cbegin` and `cend` please 😄

File: io/src/pcd_io.cpp [hunk 2]
```diff
@@ -447,6 +451,14 @@ pcl::PCDReader::readBodyASCII (std::istream &fs, pcl::PCLPointCloud2 &cloud, int
       boost::trim (line);
       boost::split (st, line, boost::is_any_of ("\t\r "), boost::token_compress_on);
 
+      if (st.size () != elems_per_line) // If this is not checked, an exception might occur when accessing st
```
Comment (kunaltyagi) [inline io/src/pcd_io.cpp:454]: might occur **on/while** accessing

File: io/src/pcd_io.cpp [hunk 3]
```diff
@@ -447,6 +451,14 @@ pcl::PCDReader::readBodyASCII (std::istream &fs, pcl::PCLPointCloud2 &cloud, int
...
+      {
+        PCL_WARN ("[pcl::PCDReader::readBodyASCII] Possibly misformed PCD file: line has %zu elements, but should have %u! (point %u)\n",
```
Comment (kunaltyagi) [inline io/src/pcd_io.cpp:456]: I think malformed is a better word in this case

Comment (kunaltyagi) [reply io/src/pcd_io.cpp:456]: Rewording the last sentence might also help: `Point number %u has %zu elements, but should have %u` or similar

**Review by kunaltyagi — APPROVED** [review]

**Pull request merged**